compiler/rustc_mir/src/transform/remove_zsts.rs

```diff
@@ -8,9 +8,6 @@ pub struct RemoveZsts;
 
 impl<'tcx> MirPass<'tcx> for RemoveZsts {
     fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
-        if tcx.sess.mir_opt_level() < 3 {
```

Comment:
Should this be kept and use `< 1` instead? I noticed that some const errors have changed, so I'm wondering if it's being triggered on `mir-opt-level=0`.

Reply:
It doesn't run at mir-opt-level=0.

See https://github.com/rust-lang/rust/blob/79e5814f4520f2c51b5307421db45cd82d134e76/compiler/rustc_mir/src/transform/mod.rs#L498 and https://github.com/rust-lang/rust/blob/79e5814f4520f2c51b5307421db45cd82d134e76/compiler/rustc_mir/src/transform/mod.rs#L558

The error changes because it removes an assignment, so we hit a different error path due to a different use of the constant. The specific error messages changed in this PR are fine (see https://github.com/rust-lang/rust/pull/83177#discussion_r595942238).

But this area seems undertested, I think I've found some cases where it's wrong, I'll open another PR that we should merge before this one.

Reply:
> The error changes because it removes an assignment, so we hit a different error path due to a different use of the constant. The specific error messages changed in this PR are fine (see [#83177 (comment)](https://github.com/rust-lang/rust/pull/83177#discussion_r595942238)).

I see, does the mir-opt-level apply to const-eval as well?

Reply:
I believe so, because I believe the same MIR is used for const eval and codegen.

But I don't think that's relevant here--the error that changes is in `main`, not in const eval'd code.

There is a separate const eval error due to the panic (reported as a future compat warning): https://github.com/rust-lang/rust/blob/79e5814f4520f2c51b5307421db45cd82d134e76/src/test/ui/consts/const-eval/panic-never-type.rs#L8-L10

Reply:
Since #78407, CTFE uses unoptimized MIR to avoid situations where an undefined behaviour is hidden because of MIR optimizations.

Reply:
> But I don't think that's relevant here--the error that changes is in `main`, not in const eval'd code.

Ah, I didn't notice that. Thanks for clearing up my confusion :)

Reply:
The reason for the change in diagnostics is that const propagation now doesn't see the constant anymore, but codegen will still evaluate it because it was used in the initial MIR body (see the https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/transform/required_consts/index.html field)

Reply:
Another way to resolve that these tests get changed would be to eagerly evaluate all `required_consts` in const propagation

To stop depending on optimizations reporting these entirely, we could also evaluate the constant in https://github.com/rust-lang/rust/blob/372afcf93bf60e1a9334b107cc3d72f1b0a4b1f4/compiler/rustc_mir/src/transform/required_consts.rs#L18 and only put it into `required_consts` if the evaluation did not succeed